README.md
# Kotlin Handler USSD API

[![Platform](https://img.shields.io/badge/platform-android-brightgreen.svg)](https://developer.android.com/index.html)
[![API](https://img.shields.io/badge/API-23%2B-brightgreen.svg?style=flat)](https://android-arsenal.com/api?level=23)
[![License](https://img.shields.io/badge/license-Apache%202.0-blue.svg)](https://github.com/romellfudi/KotlinVoIpUSSD/blob/master/LICENSE)
[![Bintray](https://img.shields.io/bintray/v/romllz489/maven/kotlin-ussd-library.svg)](https://bintray.com/romllz489/maven/kotlin-ussd-library)
[![Android Arsenal]( https://img.shields.io/badge/Android%20Arsenal-Kotlin%20Void%20USSD%20Library-green.svg?style=flat )]( https://android-arsenal.com/details/1/7628 )
[![Jitpack](https://jitpack.io/v/romellfudi/KotlinVoIpUSSD.svg)](https://jitpack.io/#romellfudi/KotlinVoIpUSSD)
[![CircleCi](https://img.shields.io/circleci/project/github/romellfudi/KotlinVoIpUSSD.svg)](https://circleci.com/gh/romellfudi/KotlinVoIpUSSD/tree/master)
[![](https://img.shields.io/badge/language-ES-blue.svg)](./README.es)

### by Romell Dominguez
[![](https://raw.githubusercontent.com/romellfudi/assets/master/favicon.ico)](https://www.romellfudi.com/)

## Target Development [High Quality](https://raw.githubusercontent.com/romellfudi/KotlinVoIpUSSD/Rev04/snapshot/device_recored.gif):

![](snapshot/device_recored.gif#gif)

Interactive with ussd windoows, It is necessary to have present that the interface depends on the SO and on the manufacturer of Android device.

## USSD KOTLIN LIBRARY 

`latestVersion` is [![Bintray](https://img.shields.io/bintray/v/romllz489/maven/kotlin-ussd-library.svg)](https://bintray.com/romllz489/maven/kotlin-ussd-library)

Add the following in your app's `build.gradle` file:

```groovy
repositories {
    jcenter()
}
dependencies {
    implementation 'com.romellfudi.ussdlibrary:kotlin-ussd-library:{latestVersion}'
}
```

* Writing xml config file from [here](https://github.com/romellfudi/KotlinVoIpUSSD/blob/master/ussd-library/src/main/res/xml/ussd_service.xml) to res/xml folder (if necessary), this config file allow link between App and SO:

```xml
<?xml version="1.0" encoding="utf-8"?>
<accessibility-service xmlns:android="http://schemas.android.com/apk/res/android"
    .../>
```

### Application

Puts dependencies on manifest, into manifest put CALL_PHONE, READ_PHONE_STATE and SYSTEM_ALERT_WINDOW:

```xml
    <uses-permission android:name="android.permission.CALL_PHONE" />
    <uses-permission android:name="android.permission.SYSTEM_ALERT_WINDOW" />
    <uses-permission android:name="android.permission.READ_PHONE_STATE" />
```

Add service:

```xml
    <service
        android:name="com.romellfudi.ussdlibrary.USSDService"
        android:permission="android.permission.BIND_ACCESSIBILITY_SERVICE">
        <intent-filter>
            <action android:name="android.accessibilityservice.AccessibilityService" />
        </intent-filter>
        <meta-data
            android:name="android.accessibilityservice"
            android:resource="@xml/ussd_service" />
    </service>
```

# How use:

First you need an hashMap from detect witch USSD' response contains the login and error messages

| KEY MESSAGE | String Messages |
| ------ | ------ |
| KEY_LOGIN | "espere","waiting","loading","esperando",... |
| KEY_ERROR | "problema","problem","error","null",... |

```kotlin
map = HashMap()
map!!["KEY_LOGIN"] = HashSet(Arrays.asList("espere", "waiting", "loading", "esperando"))
map!!["KEY_ERROR"] = HashSet(Arrays.asList("problema", "problem", "error", "null"))
```

Instance an object ussController with context

```kotlin
USSDApi ussdApi = USSDController.getInstance(activity!!)
ussdApi.callUSSDOverlayInvoke(phoneNumber, map!!, object : USSDController.CallbackInvoke {
    override fun responseInvoke(message: String) {
        // message has the response string data
        var dataToSend = "data"// <- send "data" into USSD's input text
        ussdApi!!.send("1") { // it: response String
                // message has the response string data from USSD
        }
     }

    override fun over(message: String) {
        // message has the response string data from USSD or error
        // response no have input text, NOT SEND ANY DATA
    }
})
```

if you need work with your custom messages, use this structure:

```kotlin
ussdApi.callUSSDOverlayInvoke(phoneNumber, map!!, object : USSDController.CallbackInvoke {
    override fun responseInvoke(message: String) {
        // first option list - select option 1
        ussdApi!!.send("1") { // it: response response
                // second option list - select option 1
                ussdApi.send("1") { // it: response message
                        ...
                }
        }
    }

    override fun over(message: String) {
        // message has the response string data from USSD
        // response no have input text, NOT SEND ANY DATA
    }
    ...
})
```

## Static Methods
In case use at android >= M, you could check previusly permissions, `callInvoke` and `callUSSDOverlayInvoke` methods check eneble too:

```kotlin
 // check if accessibility permissions is enable
    USSDController.verifyAccesibilityAccess(Activity)
 // check if overlay permissions is enable
    USSDController.verifyOverLay(Activity)
```

for dual sim support

```kotlin
ussdApi.callUSSDOverlayInvoke(phoneNumber, simSlot, map!!, object : USSDController.CallbackInvoke {
    ...
}
```

## Overlay Service Widget (not required)

A huge problem working with ussd is you can not invisible, disenable, resize or put on back in progressDialog
But now on Android O, Google allow build a nw kind permission from overlay widget, my solution was a widget call OverlayShowingService:
For use need add permissions at AndroidManifest:

```xml
<uses-permission android:name="android.permission.ACTION_MANAGE_OVERLAY_PERMISSION" />
```

Using the library you could use two ways:

### SplashLoadingService

Add Broadcast Service:

```xml
<service android:name="com.romellfudi.ussdlibrary.SplashLoadingService"
         android:exported="false" />
```

Invoke like a normal service:

```kotlin
val svc = Intent(activity, OverlayShowingService::class.java)
// show layout
activity.startService(svc)
ussdApi.callUSSDOverlayInvoke(phoneNumber, map!!, object : USSDController.CallbackInvoke {
        ...
        // dismiss layout
        activity.stopService(svc)
        ...
}
```

![](snapshot/device_splash.gif#gif)

### EXTRA: Use Voip line

In this section leave the lines to call to Telcom (ussd number) for connected it:

```kotlin
ussdPhoneNumber = ussdPhoneNumber.replace("#", uri)
val uriPhone = Uri.parse("tel:$ussdPhoneNumber")
context.startActivity(Intent(Intent.ACTION_CALL, uri))
```

Once initialized the call will begin to receive and send the **famous USSD windows**

![image](snapshot/telcom.png#center)

### License
```
Copyright 2019 Romell D.Z.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```

<style>
img[src*='#center'] { 
    width:390px;
    display: block;
    margin: auto;
}
img[src*='#gif'] { 
    width:200px;
    display: block;
    margin: auto;
}
</style>

CHANGELOG.md
# Changelog

## Release Kotlin USSD Library 1.1.i (19/04/2019)
- Kotlin support (kotlin-ussd-librrary)

latestVersion is 1.1.i

repositories {
    jcenter()
}
dependencies {
    implementation 'com.romellfudi.ussdlibrary:kotlin-ussd-library:{latestVersion}'
}
---

## Release USSD Library 1.1.i Interface for usage (06/03/2019)
- Overlay permissions moves into library and have static methods
---

## Release USSD Library 1.1.h Interface for usage (13/02/2019)
- Avoid issues when uses the reference: use USSDApi instead USSDController
---

## Release USSD Library 1.1.g Support multi-slot (23/01/2019)
- Release support multiple sim slots & increase minSDK to 23
---

## Release USSD Library 1.1.e Support for ITEL devices (21/01/2019)
- Release context instead of activity & support for itel devices
---

## Release USSD Library 1.1.d Splash Overlay (27/10/2018)
- Release adding second Overlay using full device screen, don't show any flashing window
---

## Release USSD Library 1.1.c Mapping Message (04/10/2018)
- Release using mapping from detect login and error pages
---

## Release USSD Android Library 1.1.b (27/09/2018)
New Structure of Messaging, reading and sending for each message

latestVersion is 1.1.b

repositories {
    jcenter()
}
dependencies {
    implementation 'com.romellfudi.ussdlibrary:ussd-library:{latestVersion}'
}
---

## Release upload on Maven Central (15/09/2018)
Upload on Maven Central

repositories {
    jcenter()
}
dependencies {
    compile 'com.romellfudi.ussdlibrary:ussd-library:{latestVersion}'
}
---

## Release USSD Android Library 1.0.b (27/08/2018)
Upgrade con.romellfudi for Android Library.

- Import ussdlibrary-1.0.b.aar
- put ussd_service.xml into res/xml folder

or import into project:

maven {
        url 'https://dl.bintray.com/romllz489/maven/'
}
dependencies {
    compile 'com.romellfudi.ussdlibrary:ussd-library:{latestVersion}'
}
---

## Release USSD Android Library 1.0.a (31/07/2018)
Remember import the internal xml into aar or replace with xml similar to ussd_service.xml

app/src/main/java/com/romellfudi/ussd/sample/MainFragment.kt
package com.romellfudi.ussd.sample

import android.Manifest.permission
import android.annotation.SuppressLint
import android.content.Context.TELEPHONY_SERVICE
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.support.v4.app.ActivityCompat
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatActivity
import android.telephony.TelephonyManager
import android.telephony.TelephonyManager.*
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast

import com.romellfudi.permission.PermissionService
import com.romellfudi.ussd.R
import com.romellfudi.ussdlibrary.OverlayShowingService
import com.romellfudi.ussdlibrary.SplashLoadingService
import com.romellfudi.ussdlibrary.USSDApi
import com.romellfudi.ussdlibrary.USSDController

import java.util.ArrayList
import java.util.Arrays
import java.util.HashMap
import java.util.HashSet

/**
 * Use Case for Test Windows
 *
 * @author Romell Domínguez
 * @version 1.1.b 27/09/2018
 * @since 1.0.a
 */
class MainFragment : Fragment() {

    private var result: TextView? = null
    private var phone: EditText? = null
    private var btn1: Button? = null
    private var btn2: Button? = null
    private var btn3: Button? = null
    private var btn4: Button? = null
    private var map: HashMap<String, HashSet<String>>? = null
    private var ussdApi: USSDApi? = null
    private var menuActivity: MainActivity? = null

    private val callback = object : PermissionService.Callback() {
        override fun onRefuse(RefusePermissions: ArrayList<String>) {
            Toast.makeText(context,
                    getString(R.string.refuse_permissions),
                    Toast.LENGTH_SHORT).show()
            activity!!.finish()
        }

        override fun onFinally() {
            // pass
        }
    }

    @SuppressLint("MissingPermission")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        map = HashMap()
        map!!["KEY_LOGIN"] = HashSet(Arrays.asList("espere", "waiting", "loading", "esperando"))
        map!!["KEY_ERROR"] = HashSet(Arrays.asList("problema", "problem", "error", "null"))
        ussdApi = USSDController.getInstance(activity!!)
        menuActivity = activity as MainActivity?
        PermissionService(activity).request(
                arrayOf(permission.CALL_PHONE, permission.READ_PHONE_STATE),
                callback)


//        val telephonyManager =(activity as MainActivity)?.getSystemService(TELEPHONY_SERVICE) as TelephonyManager
//        val handler = Handler()
//
//        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
//            Log.d("APP", "onReceiveUssdResponse")
//            telephonyManager.sendUssdRequest("*515#",
//                    object : UssdResponseCallback() {
//                        override fun onReceiveUssdResponse(telephonyManager: TelephonyManager?, request: String?, response: CharSequence?) {
//                            Log.d("APP", "onReceiveUssdResponse-request-${request}")
//                            Log.d("APP", "onReceiveUssdResponse-response-${response}")
//                            super.onReceiveUssdResponse(telephonyManager, request, response)
//                        }
//
//                        override fun onReceiveUssdResponseFailed(telephonyManager: TelephonyManager?, request: String?, failureCode: Int) {
//                            Log.d("APP", "onReceiveUssdResponseFailed-${request}")
//                            super.onReceiveUssdResponseFailed(telephonyManager, request, failureCode)
//                        }
//            },handler)
//        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.content_op1, container, false)
        result = view.findViewById<View>(R.id.result) as TextView
        phone = view.findViewById<View>(R.id.phone) as EditText
        btn1 = view.findViewById<View>(R.id.btn1) as Button
        btn2 = view.findViewById<View>(R.id.btn2) as Button
        btn3 = view.findViewById<View>(R.id.btn3) as Button
        btn4 = view.findViewById<View>(R.id.btn4) as Button
        setHasOptionsMenu(false)

        btn1!!.setOnClickListener {
            val phoneNumber = phone!!.text.toString().trim { it <= ' ' }
            ussdApi = USSDController.getInstance(activity!!)
            result!!.text = ""
            ussdApi!!.callUSSDInvoke(phoneNumber, map!!, object : USSDController.CallbackInvoke {
                override fun responseInvoke(message: String) {
                    Log.d("APP", message)
                    result!!.append("\n-\n$message")
                    // first option list - select option 1
                    ussdApi!!.send("1", object : USSDController.CallbackMessage {
                        override fun responseMessage(message: String) {
                            Log.d("APP", message)
                            result!!.append("\n-\n$message")
                            // second option list - select option 1
                            ussdApi!!.send("1", object : USSDController.CallbackMessage {
                                override fun responseMessage(message: String) {
                                    Log.d("APP", message)
                                    result!!.append("\n-\n$message")
                                }
                            })
                        }
                    })
                }

                override fun over(message: String) {
                    Log.d("APP", message)
                    result!!.append("\n-\n$message")
                }
            })
        }

        btn2!!.setOnClickListener {
            if (USSDController.verifyOverLay(activity!!)) {
                val svc = Intent(activity, OverlayShowingService::class.java)
                svc.putExtra(OverlayShowingService.EXTRA, "PROCESANDO")
                activity!!.startService(svc)
                Log.d("APP", "START OVERLAY DIALOG")
                val phoneNumber = phone!!.text.toString().trim { it <= ' ' }
                ussdApi = USSDController.getInstance(activity!!)
                result!!.text = ""
                ussdApi!!.callUSSDOverlayInvoke(phoneNumber, map!!, object : USSDController.CallbackInvoke {
                    override fun responseInvoke(message: String) {
                        Log.d("APP", message)
                        result!!.append("\n-\n$message")
                        // first option list - select option 1
                        ussdApi!!.send("1", object : USSDController.CallbackMessage {
                            override fun responseMessage(message: String) {
                                Log.d("APP", message)
                                result!!.append("\n-\n$message")
                                // second option list - select option 1
                                ussdApi!!.send("1", object : USSDController.CallbackMessage {
                                    override fun responseMessage(message: String) {
                                        Log.d("APP", message)
                                        result!!.append("\n-\n$message")
                                        activity!!.stopService(svc)
                                        Log.d("APP", "STOP OVERLAY DIALOG")
                                        Log.d("APP", "successful")
                                    }
                                })
                            }
                        })
                    }

                    override fun over(message: String) {
                        Log.d("APP", message)
                        result!!.append("\n-\n$message")
                        activity!!.stopService(svc)
                        Log.d("APP", "STOP OVERLAY DIALOG")
                    }
                })
            }
        }

        btn4!!.setOnClickListener {
            if (USSDController.verifyOverLay(activity!!)) {
                val svc = Intent(activity, SplashLoadingService::class.java)
                activity!!.startService(svc)
                Log.d("APP", "START SPLASH DIALOG")
                val phoneNumber = phone!!.text.toString().trim { it <= ' ' }
                result!!.text = ""
                ussdApi!!.callUSSDOverlayInvoke(phoneNumber, map!!, object : USSDController.CallbackInvoke {
                    override fun responseInvoke(message: String) {
                        Log.d("APP", message)
                        result!!.append("\n-\n$message")
                        // first option list - select option 1
                        ussdApi!!.send("1", object : USSDController.CallbackMessage {
                            override fun responseMessage(message: String) {
                                Log.d("APP", message)
                                result!!.append("\n-\n$message")
                                // second option list - select option 1
                                ussdApi!!.send("1", object : USSDController.CallbackMessage {
                                    override fun responseMessage(message: String) {
                                        Log.d("APP", message)
                                        result!!.append("\n-\n$message")
                                        activity!!.stopService(svc)
                                        Log.d("APP", "STOP SPLASH DIALOG")
                                        Log.d("APP", "successful")
                                    }
                                })
                            }
                        })
                    }

                    override fun over(message: String) {
                        Log.d("APP", message)
                        result!!.append("\n-\n$message")
                        activity!!.stopService(svc)
                        Log.d("APP", "STOP OVERLAY DIALOG")
                    }
                })
            }
        }

        btn3!!.setOnClickListener { USSDController.verifyAccesibilityAccess(activity!!) }

        return view
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        callback.handler(permissions, grantResults)
    }
}

